File: examples3d/debug_sleeping_kinematic3.rs
use rapier_testbed3d::TestbedViewer;
use rapier3d::prelude::*;

pub async fn run(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Setup a velocity-based kinematic rigid body.
     */
    let platform_body =
        RigidBodyBuilder::kinematic_velocity_based().translation(Vector::new(0.0, 1.5 + 0.8, 0.0));
    let collider = ColliderBuilder::cuboid(5.0, 0.5, 5.0);
    let (platform_handle, _) = world.insert(platform_body, collider);

    // A second velocity-based platform but this one will move super slow.
    let slow_platform_body =
        RigidBodyBuilder::kinematic_velocity_based().translation(Vector::new(0.0, 0.0, 0.0));
    let collider = ColliderBuilder::cuboid(5.0, 0.5, 5.0);
    let (slow_platform_handle, _) = world.insert(slow_platform_body, collider);

    /*
     * Setup the platform control parameters.
     */
    let start_tick = 500;
    let stop_tick = 1000;

    /*
     * Run the simulation.
     */
    viewer.set_world(&mut world);
    viewer.look_at(Vec3::new(10.0, 5.0, 10.0), Vec3::ZERO);

    let mut step_id = 0usize;
    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            world.step();
            step_id += 1;

            let timestep_id = step_id;
            let time = step_id as f32 * world.integration_parameters.dt as f32;

            if timestep_id == stop_tick {
                println!("Both platforms should stop moving now and eventually fall asleep.");
                // The platforms moved until this time. They must not be sleeping.
                assert!(!world.bodies[platform_handle].is_sleeping());
                assert!(!world.bodies[slow_platform_handle].is_sleeping());

                if let Some(slow_platform) = world.bodies.get_mut(slow_platform_handle) {
                    slow_platform.set_linvel(Vector::ZERO, true);
                }
                if let Some(platform) = world.bodies.get_mut(platform_handle) {
                    platform.set_linvel(Vector::ZERO, true);
                }
            }

            if timestep_id > stop_tick + 500 {
                // Platforms should fall asleep shortly after we stopped moving them.
                assert!(world.bodies[platform_handle].is_sleeping());
                assert!(world.bodies[slow_platform_handle].is_sleeping());
            }

            if !(timestep_id < start_tick || timestep_id >= stop_tick) {
                if timestep_id == start_tick {
                    println!("Platforms should start moving now and never stop.");
                    println!("The slow platform should move up and not sleep.");
                    // Platforms should have had plenty of time to fall asleep before we start moving them.
                    assert!(world.bodies[platform_handle].is_sleeping());
                    assert!(world.bodies[slow_platform_handle].is_sleeping());

                    let slow_velocity = Vector::new(0.0, 0.01, 0.0);
                    if let Some(slow_platform) = world.bodies.get_mut(slow_platform_handle) {
                        slow_platform.set_linvel(slow_velocity, true);
                    }
                }

                let velocity = Vector::new(0.0, (time * 2.0).cos(), time.sin() * 2.0);

                // Update the velocity-based kinematic body by setting its velocity.
                if let Some(platform) = world.bodies.get_mut(platform_handle) {
                    platform.set_linvel(velocity, true);
                }
            }
        }
    }
    Ok(())
}

File: python/rapier-testbed/rapier_testbed/examples3/debug_sleeping_kinematic3.py
"""Port of ``examples3d/debug_sleeping_kinematic3.rs``.

Two velocity-based kinematic platforms. From step ``start_tick=500`` they
start moving (one fast, one very slow); at ``stop_tick=1000`` both stop
and should eventually fall asleep.
"""
from __future__ import annotations

import math

import rapier3d as rp

from .._registry import register

CATEGORY = "Debug"
NAME = "Sleeping kinematics"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    plat_body = rp.RigidBody.kinematic_velocity_based().translation(
        (0.0, 1.5 + 0.8, 0.0)
    )
    plat_handle = bodies.insert(plat_body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(5.0, 0.5, 5.0), plat_handle, bodies
    )

    slow_body = rp.RigidBody.kinematic_velocity_based().translation((0.0, 0.0, 0.0))
    slow_handle = bodies.insert(slow_body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(5.0, 0.5, 5.0), slow_handle, bodies
    )

    start_tick = 500
    stop_tick = 1000

    def _cb(tb) -> None:
        step = tb._step_count
        if step == stop_tick:
            for h in (plat_handle, slow_handle):
                b = tb.bodies.get(h)
                if b is not None:
                    b.linvel = rp.Vec3(0.0, 0.0, 0.0)
        if step < start_tick or step >= stop_tick:
            return
        if step == start_tick:
            slow_velocity = rp.Vec3(0.0, 0.01, 0.0)
            slow = tb.bodies.get(slow_handle)
            if slow is not None:
                slow.linvel = slow_velocity
        t = step * tb._integration_parameters.dt
        velocity = rp.Vec3(0.0, math.cos(t * 2.0), math.sin(t) * 2.0)
        plat = tb.bodies.get(plat_handle)
        if plat is not None:
            plat.linvel = velocity

    testbed.add_callback(_cb)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((10.0, 5.0, 10.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
